modules/tinymce/src/themes/silver/main/ts/ui/skin/Loader.ts

```diff
@@ -40,14 +40,71 @@ const loadSkin = (isInline: boolean, editor: Editor): Promise<void> => {
   // Seems to work without, but adding a note in case things break later
   if (!Options.isSkinDisabled(editor) && Type.isString(skinUrl)) {
     return Promise.all([
-      loadUiSkins(editor, skinUrl),
-      loadShadowDomUiSkins(editor, skinUrl)
-    ]).then(SkinLoaded.fireSkinLoaded(editor), SkinLoaded.fireSkinLoadError(editor, 'Skin could not be loaded'));
+      loadUrlUiSkins(editor, skinUrl),
+      loadShadowDomUrlUiSkins(editor, skinUrl)
+    ]);
   } else {
-    return Promise.resolve(SkinLoaded.fireSkinLoaded(editor)());
+    return Promise.resolve();
   }
 };
 
+const loadRawCss = (editor: Editor, key: string, css: string, styleSheetLoader: StyleSheetLoader): void => {
+  // Ensure the stylesheet is cleaned up when the editor is destroyed
+  editor.on('remove', () => styleSheetLoader.unloadRawCss(key));
+  return styleSheetLoader.loadRawCss(key, css);
+};
+
+const loadRawUiSkins = (editor: Editor, skinKey: string): void => {
+  const skinUiCss = skinKey + '/skin.css';
+  const css = EditorManager.resources.get(skinUiCss);
+
+  if (Type.isString(css)) {
+    return loadRawCss(editor, skinUiCss, css, editor.ui.styleSheetLoader);
+  }
+};
+
+const loadShadowDomRawUiSkins = (editor: Editor, skinKey: string): void => {
+  const isInShadowRoot = SugarShadowDom.isInShadowRoot(SugarElement.fromDom(editor.getElement()));
+  if (isInShadowRoot) {
+    const shadowDomSkinCss = skinKey + '/skin.shadowdom.css';
+    const css = EditorManager.resources.get(shadowDomSkinCss);
```

Comment:
This duplication makes me cringe. I wonder if we could re-use `skin_url` rather than making the new `skin_key` with all the dancing around how it's similar but different.

It's a big change, but then we'd check if the URL was in the editor resources, if it is load as raw CSS if not load as URL.

Does that make sense? Would we have time for a change that big?

Reply:
When I made this I actively did not want to reuse anything existing due to the exciting risks that can come from overloading an option.
Not saying I'm a big fan of the duplication either, so not like I'd be opposed to removing it somehow.
And the time-limit I don't know, it seems like we're moving this to... patch version? Next version? I think?

Reply:
I think it's a valid risk to take - it would also allow developers to cache their custom resources by simply adding the contents with the full URL as the key.